pilot/pkg/secrets/kube/secrets.go

```diff
@@ -47,14 +56,139 @@ const (
-	return &SecretsController{informer}
+	_ = client.KubeInformer().Core().V1().Secrets().Informer()
+
+	return &SecretsController{
+		secrets: client.KubeInformer().Core().V1().Secrets(),
+
+		authorizationEnabled: true,
+		sar:                  client.AuthorizationV1().SubjectAccessReviews(),
+		clusterID:            clusterID,
+		remoteGetter:         remoteClientGetter,
+		authorizationCache:   make(map[authorizationKey]authorizationResponse),
+	}
+}
+
+func toUser(serviceAccount, namespace string) string {
+	return fmt.Sprintf("system:serviceaccount:%s:%s", namespace, serviceAccount)
+}
+
+const cacheTTL = time.Minute
+
+// cachedAuthorization checks the authorization cache
+// nolint
+func (s *SecretsController) cachedAuthorization(user, clusterID string) (error, bool) {
+	key := authorizationKey{cluster: clusterID, user: user}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.authorizationEnabled {
+		return nil, true
+	}
+	got, f := s.authorizationCache[key]
+	if !f {
+		return nil, false
+	}
+	// Response found, but its expired. Evict it.
+	if got.expiration.Before(time.Now()) {
+		delete(s.authorizationCache, key)
+		return nil, false
+	}
+	return got.authorized, true
+}
+
+// cachedAuthorization checks the authorization cache
+func (s *SecretsController) insertCache(user, clusterID string, response error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	key := authorizationKey{cluster: clusterID, user: user}
+	expDelta := cacheTTL
+	if response == nil {
+		// Cache success a bit longer, there is no need to quickly revoke access
+		expDelta *= 5
+	}
+	log.Debugf("cached authorization for user %s: %v", user, response)
+	s.authorizationCache[key] = authorizationResponse{
+		expiration: time.Now().Add(expDelta),
+		authorized: response,
+	}
+}
+
+func (s *SecretsController) getAccessReviewClient(clusterID string) authorizationv1client.SubjectAccessReviewInterface {
+	// first match local/primary cluster
+	if s.clusterID == clusterID || clusterID == "" {
+		return s.sar
+	}
+
+	// secondly try other remote clusters
+	if s.remoteGetter != nil {
+		if res := s.remoteGetter(clusterID); res != nil {
+			return res.AuthorizationV1().SubjectAccessReviews()
+		}
+	}
+
+	return nil
+}
+
+// DisableAuthorization makes the authorization check always pass. Should be used only for tests.
+func (s *SecretsController) DisableAuthorization() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.authorizationEnabled = false
+}
+
+func (s *SecretsController) Authorize(serviceAccount, namespace, clusterID string) error {
+	user := toUser(serviceAccount, namespace)
+	if cached, f := s.cachedAuthorization(user, clusterID); f {
+		return cached
+	}
+	resp := func() error {
+		resp, err := s.sar.Create(context.Background(), &authorizationv1.SubjectAccessReview{
+			ObjectMeta: metav1.ObjectMeta{},
+			Spec: authorizationv1.SubjectAccessReviewSpec{
+				ResourceAttributes: &authorizationv1.ResourceAttributes{
+					Namespace: namespace,
+					Verb:      "list",
+					Resource:  "secrets",
```

Comment:
or get? And specify group version?

Reply:
This seems to only work when I leave them empty... I think maybe `core` group is special?

Reply:
core group, the group is "", but i think the version should be set "v1"

Reply:
BTW, "get" verbs should be added

Reply:
We can only do one or the other unless we have 2 calls. Do you suggest we switch to get or that we make two calls?

Reply:
I think we should do 2 API calls to cover both cases

Reply:
I don't personally see the benefit. This is intended to be a short term approach to support existing users, who currently MUST have `list`. For longer term approach we will address extended use cases but for now it doesn't seem needed (https://docs.google.com/document/d/1nwTh2g7FgdkC8suc3SHZ5le_4Rkm8YvLY8Cgah4cm-I/edit#)

Reply:
I kind of agree that 'get' is the critical one. Yes, they now have list - so it's not a big deal, and I don't think we need to checks - but if we make only one 'get' might be better.

But since it's mainly for upgrade - I'm ok with list as well.